### ISSUE TITLE
Your work is really great! I would like to ask you about your EBGM module！

### ISSUE THREAD

**Issue by WCWCWCQAQAQA** [YoeTon/CTUNet#1]
Your improved attention module has deeply inspired my work, could you please share your EGSA code with me so that I can have a deeper understanding of your work?

**Comment by YoeTon** [comment]
你好，相关部分代码放在附件里了。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "YoeTon/CTUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年9月13日(星期五) 晚上6:43
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;[YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1)





 
Your improved attention module has deeply inspired my work, could you please share your EGSA code with me so that I can have a deeper understanding of your work?

**Comment by WCWCWCQAQAQA** [comment]
抱歉，我好像没有看到附件，您是不是不小心没点击上呀？

---- Replied Message ----
| From | ***@***.***> |
| Date | 09/17/2024 17:30 |
| To | YoeTon/CTUNet ***@***.***> |
| Cc | Rui ***@***.***>,
Author ***@***.***> |
| Subject | Re: [YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1) |

你好，相关部分代码放在附件里了。




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "YoeTon/CTUNet" ***@***.***&gt;;
发送时间:&nbsp;2024年9月13日(星期五) 晚上6:43
***@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;[YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1)






Your improved attention module has deeply inspired my work, could you please share your EGSA code with me so that I can have a deeper understanding of your work?

**Comment by YoeTon** [comment]
class EGSA(nn.Module):


&nbsp; &nbsp; def __init__(self, dim, num_heads, bias):
&nbsp; &nbsp; &nbsp; &nbsp; super(EGSA, self).__init__()
&nbsp; &nbsp; &nbsp; &nbsp; self.num_heads = num_heads
&nbsp; &nbsp; &nbsp; &nbsp; self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))
&nbsp; &nbsp; &nbsp; &nbsp; self.conv0 = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)
&nbsp; &nbsp; &nbsp; &nbsp; self.edge_body = edge(dim)
&nbsp; &nbsp; &nbsp; &nbsp; self.conv1 = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)
&nbsp; &nbsp; &nbsp; &nbsp; self.project_out = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)
&nbsp; &nbsp; &nbsp; &nbsp; self.relu = nn.ReLU(inplace=False)


&nbsp; &nbsp; def forward(self, x):
&nbsp; &nbsp; &nbsp; &nbsp; b, c, h, w = x.shape
&nbsp; &nbsp; &nbsp; &nbsp; x_edge = self.edge_body(x)
&nbsp; &nbsp; &nbsp; &nbsp; x_q = x + 0.01 * x_edge
&nbsp; &nbsp; &nbsp; &nbsp; q = self.conv0(x_q)
&nbsp; &nbsp; &nbsp; &nbsp; k = q
&nbsp; &nbsp; &nbsp; &nbsp; v = self.conv1(x)


&nbsp; &nbsp; &nbsp; &nbsp; q = rearrange(q, 'b (head c) h w -&gt; b head c (h w)', head=self.num_heads)
&nbsp; &nbsp; &nbsp; &nbsp; k = rearrange(k, 'b (head c) h w -&gt; b head c (h w)', head=self.num_heads)
&nbsp; &nbsp; &nbsp; &nbsp; v = rearrange(v, 'b (head c) h w -&gt; b head c (h w)', head=self.num_heads)
&nbsp; &nbsp; &nbsp; &nbsp; q = torch.nn.functional.normalize(q, dim=-1)
&nbsp; &nbsp; &nbsp; &nbsp; k = torch.nn.functional.normalize(k, dim=-1)


&nbsp; &nbsp; &nbsp; &nbsp; attn = (q @ k.transpose(-2, -1)) * self.temperature
&nbsp; &nbsp; &nbsp; &nbsp; attn = attn.softmax(dim=-1)


&nbsp; &nbsp; &nbsp; &nbsp; out = (attn @ v)


&nbsp; &nbsp; &nbsp; &nbsp; out = rearrange(out, 'b head c (h w) -&gt; b (head c) h w', head=self.num_heads, h=h, w=w)
&nbsp; &nbsp; &nbsp; &nbsp; out = self.project_out(out)
&nbsp; &nbsp; &nbsp; &nbsp; return out


class edge(nn.Module):
&nbsp; &nbsp; def __init__(self, n_feats):
&nbsp; &nbsp; &nbsp; &nbsp; super(edge, self).__init__()
&nbsp; &nbsp; &nbsp; &nbsp; self.conv = nn.Conv2d(n_feats * 2, 2, kernel_size=(3, 3), stride=1, padding=(1, 1))


&nbsp; &nbsp; def forward(self, out):
&nbsp; &nbsp; &nbsp; &nbsp; size = out.size()[2:]
&nbsp; &nbsp; &nbsp; &nbsp; flow = F.interpolate(out, (int(size[0] / 4.), int(size[1] / 4.)), mode='bilinear', align_corners=False)
&nbsp; &nbsp; &nbsp; &nbsp; flow = F.interpolate(flow, size, mode='bilinear', align_corners=False)
&nbsp; &nbsp; &nbsp; &nbsp; flow = self.conv(torch.cat([out, flow], 1))
&nbsp; &nbsp; &nbsp; &nbsp; flow = flow_warp(out, flow, size)


&nbsp; &nbsp; &nbsp; &nbsp; return out - flow


def flow_warp(input, flow, size):
&nbsp; &nbsp; out_h, out_w = size


&nbsp; &nbsp; n, c, h, w = input.size()


&nbsp; &nbsp; norm = torch.tensor([[[[out_w, out_h]]]]).type_as(input).to(input.device)


&nbsp; &nbsp; h_grid = torch.linspace(-1.0, 1.0, out_h).view(-1, 1).repeat(1, out_w)
&nbsp; &nbsp; w_gird = torch.linspace(-1.0, 1.0, out_w).repeat(out_h, 1)
&nbsp; &nbsp; grid = torch.cat((w_gird.unsqueeze(2), h_grid.unsqueeze(2)), 2)


&nbsp; &nbsp; grid = grid.repeat(n, 1, 1, 1).type_as(input).to(input.device)
&nbsp; &nbsp; grid = grid + flow.permute(0, 2, 3, 1) / norm


&nbsp; &nbsp; output = F.grid_sample(input, grid, align_corners=False)


&nbsp; &nbsp; return output



---Original---
From: "Rui ***@***.***&gt;
Date: Tue, Sep 17, 2024 18:09 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1)




 抱歉，我好像没有看到附件，您是不是不小心没点击上呀？
 
 ---- Replied Message ----
 | From | ***@***.***&gt; |
 | Date | 09/17/2024 17:30 |
 | To | YoeTon/CTUNet ***@***.***&gt; |
 | Cc | Rui ***@***.***&gt;,
 Author ***@***.***&gt; |
 | Subject | Re: [YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1) |
 
 你好，相关部分代码放在附件里了。
 
 
 
 
 ------------------&amp;nbsp;原始邮件&amp;nbsp;------------------
 发件人: "YoeTon/CTUNet" ***@***.***&amp;gt;;
 发送时间:&amp;nbsp;2024年9月13日(星期五) 晚上6:43
 ***@***.***&amp;gt;;
 ***@***.***&amp;gt;;
 主题:&amp;nbsp;[YoeTon/CTUNet] Your work is really great! I would like to ask you about your EBGM module！ (Issue #1)
 
 
 
 
 
 
 Your improved attention module has deeply inspired my work, could you please share your EGSA code with me so that I can have a deeper understanding of your work?

**Comment by WCWCWCQAQAQA** [comment]
谢谢您的耐心回复！已看到！
